Documentatition (#15)
it-at-m/lhm_actions: docs/deployment.md

> ![image](images/ci_cd_github_big_picture_public.drawio.png)

Author: Ich habe doch unter versucht. In den Prozess zu erklären. Das Bild dient als Überblick.

> 

Author: done

# Helm Chart

![image](images/external-helm-chart.drawio.png)

> There is a sample chart. https://github.com/it-at-m/helm-charts/tree/main/charts/sps-sample . It consists of subcharts. Each subchart is a different module for example frontend, backend, eai.  Each module was created with `helm create` and some adaptions. The Chart.yml includes the external dependency of the apigateway. The main config is done in the values.yml of the parent chart. In the section refarch-gateway, frontend, backend you can set the config of the subchart. 

Reviewer: ```suggestion
We provide a sample chart. https://github.com/it-at-m/helm-charts/tree/main/charts/sps-sample, which consists subcharts. Each subchart is a different module for example frontend, backend, eai.  Each module was created with `helm create` and we made some adaptions. The Chart.yml includes the external dependency of the [RefArch Api Gateway Helm Chart](https://github.com/it-at-m/helm-charts/tree/main/charts/refarch-gateway). The main config is done in the values.yml of the parent chart. In the sections refarch-gateway, frontend, backend you can configure the subcharts. 
```

> There is a sample chart. https://github.com/it-at-m/helm-charts/tree/main/charts/sps-sample . It consists of subcharts. Each subchart is a different module for example frontend, backend, eai.  Each module was created with `helm create` and some adaptions. The Chart.yml includes the external dependency of the apigateway. The main config is done in the values.yml of the parent chart. In the section refarch-gateway, frontend, backend you can set the config of the subchart. 

Author: done

# Deployment

At the end of a sprint a developer builds a release in the source code project. It needs to be done for each folder frontend/backend/eai separately. Then the devops engineer needs to update the chart-version and app-version of the helm chart in the Chart.yaml.

> Internal the quay syncs the docker images with the external github image registry. The github api has a pull limit so we need to cache the images internally.

Reviewer: You don't explain quay. I would prefer to keep this general and not use the tool name, but e.g. "internal Image repository". Also I would start with the 2nd sentence (why we are synching and the, how the synch ist done)

![image](images/internal-chart.drawio.png)

The internal gitlab repo you can config your application, for example the trust store, ingress (openshift route), single sign on, database config, environment configs. The most config is comment out, so you need to comment it in. Here you see an example https://git.muenchen.de/ccse/cicd/sps-github

> You have two possibilities for the image rollout. 

Author: done

> -	For the dev environment, you want to have an auto rollout. We create a image stream. The image stream has a link to the quay. In the deployment there is a special annotation set. Openshift then triggers automatically trigger a new rollout if a new image is available.  https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/images/triggering-updates-on-imagestream-changes#triggering-updates-on-imagestream-changes

Reviewer: Consider to change Quay with "internal image repository", as well replacing "Openshift" with "internal kubernetes platform".

Author: ```suggestion
-	For the dev environment, you want to have an auto rollout. We create a image stream. The image stream has a link to the internal image registry (Quay). In the deployment there is a special annotation set. The internal kubernetes platform (Openshift) then triggers automatically trigger a new rollout if a new image is available.  https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/images/triggering-updates-on-imagestream-changes#triggering-updates-on-imagestream-changes
```

Author: wenn wir jetzt in Github zur Quay "internal image repository" und Openshift (internal kubernetes plattform) nennen, kennt sich keiner mehr aus. Ich finde wir sollten für externe Doku dieselben Begriffe nehmen

Reviewer: Gerne Quay und Openshift in Klammern. Aber wenn du nur Quay und Openshift verwendest, ist es nicht mehr allgemeingültig für alle (auch Nicht-LHMler) und wenn wir intern was ändern, musst du u.U. nur deswegen die Doku ändern.

> -	For the test, prod environment, you want to manual rollout. You need to specify the image version for each service in the values.yml.  You need to pay attention that you take the image from the internal quay and not from the github registry.

Reviewer: ```suggestion
-	For the other environment (test, prod), a manual rollout is implemented. You have to specify the image version for each service in the values.yml as well paying attention to use the image from the internal image registry and not from the GitHub Registry.
```

# Architecture-Overview

![architecture-overview](images/ci_cd_github_big_picture_public.drawio.png)
_The picture shows the architecture overview. The steps are explaind in the following passages_

# Source Repo

> The top left from the architecture-overview is explaint in [link](./workflows.md).

Reviewer: ```suggestion
Execute GitHub Actions to compile and build code, build images. For more information see [workflows.md](./workflows.md).
```

# Helm Chart

![architekture-external-helm-chart](images/external-helm-chart.drawio.png)

> The top right from the architecture-overview is explaint in the following.

Reviewer: würde das streichen.

Author: ```suggestion
The top right from the architecture-overview is explaint in the following.
```

We provide a sample chart. https://github.com/it-at-m/helm-charts/tree/main/charts/sps-sample, which consists subcharts. Each subchart is a different module for example frontend, backend, eai. Each module was created with `helm create` and we made some adaptions. The Chart.yml includes the external dependency of the [RefArch Api Gateway Helm Chart](https://github.com/it-at-m/helm-charts/tree/main/charts/refarch-gateway). The main config is done in the values.yml of the parent chart. In the sections refarch-gateway, frontend, backend you can configure the subcharts.

> # Deployment

Reviewer: ```suggestion
# Internal Deployment (Image Repository `Quai`, Git Repository GitLab `git.muenchen.de`, Kubernetes platform `OpenShift`)
```

> The bottom part from the architecture-overview is expliant in the following.

Reviewer: würde ich streichen

Author: https://github.com/it-at-m/lhm_actions/pull/15#discussion_r1965290406 hier hast du gesagt, es ist nicht klar wo ich etwas erkläre. dann streiche ich es wieder

> At the end of a sprint a developer builds a release in the source code project. It needs to be done for each folder frontend/backend/eai separately. Then the devops engineer needs to update the chart-version and app-version of the helm chart in the Chart.yaml.

Reviewer: Ich würde diesen Teil wie folgt aufbauen:

1. Was nutzen wir intern (mit Verlinkung auf die offinziellen Links):
    1. Die Image Registry Quai, in das die Docker Images der Projekte synchronisiert werden
    2. GitLab zur Ausführung der internen Pipelines (IaC) und Konfiguration der Anwendungen
    3. OpenShift für den Betrieb der Anwendungen
 2. Dann wie es funktioniert
     1. Autorollout für latest images (nur in nicht-produktive Umgebungen)
     2. Versionierung für deployment in produktive oder produktionsnahe Umgebungen:
        1. über WF maven-build-release (setzen der Versionsnummer)
        2. Manuelles Triggern der internen Pipeline

Internal the quay syncs the docker images with the external github image registry. The github api has a pull limit so we need to cache the images internally. The configuration is done by the openshift team.

![architekture-internal-helm-chart](images/internal-chart.drawio.png)

> The internal gitlab repo you can config your application, for example the trust store, ingress (openshift route), single sign on, database config, environment configs. The most config is comment out, so you need to comment it in. Here you see an example https://git.muenchen.de/ccse/cicd/sps-github

Reviewer: Auf git.muenchen.de kann man nicht zugreifen von außen. Diese Verlinkung macht keinen Sinn.

Author: das ist halt auch blöd zwei Systeme Github/Gitlab zu Nutzen. Das Projekt auf Github zu Veröffentlichen macht auch keine Sinn, weil das interessante ist die Gitlab-Pipeline. Dann gehen wir zu Opencode.

Author: ich habe auch keine Probleme den internen Proxy-Adresse oder die Openshift API-Adresse zu veröffentlichen. Von außen kommt man da nicht drauf.

Reviewer: > das ist halt auch blöd zwei Systeme Github/Gitlab zu Nutzen. Das Projekt auf Github zu Veröffentlichen macht auch keine Sinn, weil das interessante ist die Gitlab-Pipeline. Dann gehen wir zu Opencode.

Du kannst mal Klaus fragen, ob es in der OSS-Doku irgendwo steht, warum wir extern GitHub verwenden und nicht OpenCode. Dann könnten wir darauf verweisen. Ansonsten könnten wir den Grund hier natürlich auch reinschreiben, wenn es Sinn macht.

Reviewer: > ich habe auch keine Probleme die internen Proxy-Adresse oder die Openshift API-Adresse zu veröffentlichen. Von außen kommt man da nicht drauf.

Das mag schon sein, aber ich finde Links im Internet, die dann ins Leere laufen und mir eine Fehlerseite anzeigen (DU hast da aber keinen Zugriff drauf) nervig und unprofessioniell.

Author: wenn ich jetzt was schreibe wie man es offen nutzt. dann geht es wieder intern nicht, wegen proxy, image nicht verfügbar etc.

Reviewer 2: > Du kannst mal Klaus fragen, ob es in der OSS-Doku irgendwo steht, warum wir extern GitHub verwenden und nicht OpenCode.

> Wir richten uns nach der Community und nutzen ebenso GitHub für Public Code der Landeshauptstadt München. Durch die Codeverwaltung auf GitHub bieten wir eine niederschwellige Teilhabe für alle an.

https://opensource.muenchen.de/de/repositories.html#code

Reviewer 2: > Here you see an example https://git.muenchen.de/ccse/cicd/sps-github

@ejcsid ich bin da bei dir das es blöd ist wenn man (nach extern) tote links im internet veröffentlicht. Aber in diesem Fall beschreiben wir ja wie wir den "Brückenkopf" wie freie Software und freie Container zu uns on-prem kommen.

Daher fände ich sogar @hupling Lösung vertretbar (mit einer Klammer dahinter)

Schön wäre es wenn wir ganz https://git.muenchen.de/ccse/cicd/sps-github/ raus mirroren (da wäre en mirror mal vertretbar;)

Opencode würde uns da nicht helfen, denn wir hätten immer ein internes gitlab für die interen CICD.

Reviewer: > Schön wäre es wenn wir ganz https://git.muenchen.de/ccse/cicd/sps-github/ raus mirroren (da wäre en mirror mal vertretbar;)

Dann machen wir doch das und verlinken darauf. @hupling dann kannst du in der Doku erstmal rein schreiben, dass ein Beispiel noch kommt.

5. _OpenShift `Container Application Platform`_: Internal kubernetes plattform to run applications

You can find further details in the following chapters.\_

## Source repository (GitHub Project `it-at-m/foo`)

Execute GitHub Actions to compile and build code, build images. For more information see [workflows.md](./workflows.md).

## Helm Chart (GitHub Project `it-at-m/helm-charts`)

![architekture-external-helm-chart](images/external-helm-chart.drawio.png)

In the repository [it-at-m/helm-charts](https://github.com/it-at-m/helm-charts) we provide helm charts for Docker images created in our project repos. There you can find the sample helm chart [sps-sample](https://github.com/it-at-m/helm-charts/tree/main/charts/sps-sample), which consists of subcharts.
Each subchart is a different module, for example frontend, backend, eai. Each module was created with `helm create` and we made some adaptions. The Chart.yml includes the external dependency of the [RefArch API Gateway Helm Chart](https://github.com/it-at-m/helm-charts/tree/main/charts/refarch-gateway). The main config is done in the values.yml of the parent chart.

> In the sections refarch-gateway, frontend, backend you can configure the subcharts.

Reviewer: ```suggestion
In the sections refarch-gateway, frontend, backend of the configuration file values.yaml you can configure the subcharts.
```

## Internal Deployment (Image Repository `Quay`, Git Repository GitLab `git.muenchen.de`, Kubernetes platform `OpenShift`)

### What are we using internally

> - Image Registry [Red Hat Quay](https://docs.redhat.com/de/documentation/red_hat_quay) for synchronising docker images

Reviewer: ```suggestion
- Image Registry [Red Hat Quay](https://docs.redhat.com/de/documentation/red_hat_quay) to synchronize Docker images
```

> - [Gitlab](https://docs.gitlab.com/) for the execution of internal pipelines (IaC) and configuration of the applications

Reviewer: ```suggestion
- [GitLab](https://docs.gitlab.com/) to execute internal infrastructure pipelines (IaC) and configuration of the applications
```

> - [Openshift](https://docs.redhat.com/en/documentation/openshift_container_platform) for the operations of the applications

Reviewer: ```suggestion
- [OpenShift](https://docs.redhat.com/en/documentation/openshift_container_platform), our Kubernetes platform to host the applications
```

### How it works together

#### Autorollout: For the dev environment (only not prductive enviornment)

> An automatic rollout is implemented. We create an image stream that links to internal Image Registry (Quay). The deployment includes a special annotation that prompts The internal kubernetes platform (Openshift) to automatically trigger a new rollout when a new image becomes available. For more details, see [this documentation](https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/images/triggering-updates-on-imagestream-changes#triggering-updates-on-imagestream-changes).

Reviewer: ```suggestion
An automatic rollout is implemented. We create an image stream that links to the internal image registry (Quay). The deployment includes a special annotation that prompts the internal kubernetes platform (OpenShift) to automatically trigger a new rollout when a new image becomes available. For more details, see [this documentation](https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/images/triggering-updates-on-imagestream-changes#triggering-updates-on-imagestream-changes).
```

#### manual rollout for the deployment in productive and close to productive environments

> For the other environment (test, prod), a manual rollout is implemented. You have to specify the image version for each service in the values.yml as well paying attention to use the image from the internal image registry and not from the GitHub Registry.

Reviewer: ```suggestion
For other environments (stage, prod), a manual rollout is implemented. You have to specify the image version for each service in the Helm configuration file values.yml as well paying attention to use the image from the internal image registry and not from the GitHub Registry.
```

# Deployment

> 

Author: done
